scripts/search_outcomes_config.py

```diff
@@ -0,0 +1,222 @@
+#!/usr/bin/env python
+"""Search an outcome file for configurations with given settings.
+
+Read an outcome file and report the configurations in which test_suite_config
+runs with the required settings (compilation option enabled or disabled).
+"""
+
+import argparse
+import os
+import re
+import subprocess
+from typing import Dict, FrozenSet, Iterator, List, Set
+import tempfile
+import unittest
+
+from mbedtls_framework import build_tree
+
+
+def make_regexp_for_settings(settings: List[str]) -> str:
+    """Construct a regexp matching the interesting outcome lines.
+
+    Interesting outcome lines are from test_suite_config where the given
+    setting is passing.
+
+    We assume that the elements of settings don't contain regexp special
+    characters.
+    """
+    return (r';test_suite_config[^;]*;Config: (' +
+            '|'.join(settings) +
+            r');PASS;')
+
+def run_grep(regexp: str, outcome_file: str) -> List[str]:
+    """Run grep on the outcome file and return the matching lines."""
+    env = os.environ.copy()
+    env['LC_ALL'] = 'C' # Speeds up some versions of GNU grep
+    try:
+        return subprocess.check_output(['grep', '-E', regexp, outcome_file],
+                                       encoding='ascii',
+                                       env=env).splitlines()
+    except subprocess.CalledProcessError as exn:
+        if exn.returncode == 1:
+            return [] # No results. We don't consider this an error.
+        raise
+
+OUTCOME_LINE_RE = re.compile(r'[^;]*;'
+                             r'([^;]*);'
+                             r'test_suite_config\.(?:[^;]*);'
+                             r'Config: ([^;]*);'
+                             r'PASS;')
+
+def extract_configuration_data(outcome_lines: List[str]) -> Dict[str, FrozenSet[str]]:
+    """Extract the configuration data from outcome lines.
+
+    The result maps a configuration name to the list of passing settings
+    in that configuration.
+    """
+    config_data = {} #type: Dict[str, Set[str]]
+    for line in outcome_lines:
+        m = OUTCOME_LINE_RE.match(line)
+        # make_regexp_for_settings() arranges to only return lines that
+        # should match here.
```

Comment:
Why do we need this comment here? I am not sure how it supposed to be helpful in this context. I mean when reading this code, where might be the question raised to which this is the answer?

Reply:
This comment explains why the assertion on the line below is always expected to succeed.

Reply:
Oh, I didn't realise that it was trying to explain that.

It seems to me that`make_regexp_for_settings` requires a single `;` character right before `test_suite_config`. `OUTCOME_LINE_RE` requires another `;` in the preceding string. If the string preceding `;test_suite_config` doesn't have a `;` it will match with `make_regexp_for_settings` but won't match with `OUTCOME_LINE_RE` and the assertion will fail. What am I missing here?

Reply:
We'll have an assert failure if the file doesn't have the expected format of an outcome file. I think that's reasonable for a developer-facing script.